File: 프로그래머스/unrated/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.java
class Solution {
    public int[][] solution(int n) {
        int[][] answer = new int[n][n];
        int idx = 0;
        int row = 0;
        int col = 0;
        int cnt = 1;
        
        boolean isRow = true;
        boolean isCol = true;
        
        while (cnt <= answer.length * answer.length) {
            if (isRow && isCol) {
                if (col < n - idx) {
                    answer[row][col++] = cnt++;
                } else {
                    col--;
                    row++;
                    isRow = !isRow;
                }
            } else if (isCol && !isRow) {
                if (row < n - idx) {
                    answer[row++][col] = cnt++;
                } else {
                    row--;
                    col--;
                    isCol = !isCol;
                    isRow = !isRow;
                }
            } else if (isRow && !isCol) {
                if (col > idx - 1) {
                    answer[row][col--] = cnt++;
                } else {
                    row--;
                    col++;
                    isRow = !isRow;
                    idx++;
                }
            } else if (!isCol && !isRow) {
                if (row > idx - 1) {
                    answer[row--][col] = cnt++;
                } else {
                    col++;
                    row++;
                    isCol = !isCol;
                    isRow = !isRow;
                } 
                
            }
        }
            
       
        return answer;
    }
}

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    answer = [[-1] * n for _ in range(n)]
    # 우 하 좌 상
    dx = [0, 1, 0, -1]
    dy = [1, 0, -1, 0]
    
    direction = 0
    num, x, y = 1, 0, 0
    while n * n >= num:
        answer[x][y] = num
        num += 1
        nx = x + dx[direction]
        ny = y + dy[direction]
        # 방향 전환을 해야할 순간
        # 1. 맵 벗어나는 경우
        # 2. 이미 값이 채워진 경우
        if nx < 0 or nx >= n or ny < 0 or ny >= n or answer[nx][ny] != -1:
            direction = (direction + 1) % 4
            nx = x + dx[direction]
            ny = y + dy[direction]
        x = nx
        y = ny
    return answer